Quantization refactor

lucadellalib opened the pull request:
- Add K-means model
- Add example recipe

See https://github.com/speechbrain/speechbrain/issues/2764.

TParcollet commented:
I will do a more precise review later when it's ready. The idea looks good to me. I'm wondering about having a kmeans into lobes ... maybe this is a place where the lobes/integrations folder will make sense? Or utils maybe? kmeans is sufficiently widely used to make it logical to have a utils/kmeans.py or smth like that, maybe.

mravanelli commented:
Thank you @lucadellalib for this PR! This significantly improves the quantization process and makes it more "SpeechBrain-like". Below are my comments:

- I tested the recipe, and everything runs as expected

- The quantization recipe in LJSpeech must also be updated. 
- We need to ensure that the new quantization recipe produces outputs compatible with HiFi-GAN discrete training. Have you tested this? See `recipes/LJSpeech/TTS/vocoder`
- This recipe uses `scikit-learn` as an extra dependency, which isn’t included in SpeechBrain by default. I suggest creating an extra-requirements.txt file and adding scikit-learn there. Please also document this new dependency in the README.md file.

- In the future, I will consider moving `speechbrain/lobes/models/kmeans.py` to the integration folder that @pplantinga is preparing. This aligns well with the idea of placing techniques relying on external dependencies in the integration folder.
- Please update the README.md with clear instructions for training. Include links to logs, the pre-trained model, etc.

Adel-Moumen commented on recipes/LibriSpeech/quantization/train_kmeans.py:
```diff
@@ -0,0 +1,258 @@
+            loss = (curr_cluster_centers - old_cluster_centers).norm()
+
+            # Workaround to keep consistency with standard supervised training
+            loss.requires_grad_()
```
why is it the case? Maybe we could directly fix core.py

lucadellalib replied:
SpeechBrain expects a loss with a gradient. This is a very special case, I don't think it's worth addressing it in `core.py`, as doing so might lead to silent bugs in other situations (typically the loss gradient should come from the model's parameters). I've added additional comments to clarify this.

Adel-Moumen commented:
Hey @lucadellalib, I like what you proposed. I think its much more in the spirit of what we are doing in SpeechBrain. I went through the code and saw some lines were fixes to make Brain work with kmeans (e.g. the returned loss). Can't we directly modify core.py to make the code more compatible ? I think there's some flexibility that could be improved on Brain side

pplantinga commented on recipes/LibriSpeech/quantization/train_kmeans.py:
```diff
@@ -0,0 +1,258 @@
+            if if_main_process():
+                self.checkpointer.save_and_keep_only(
+                    meta={"loss": stage_stats["loss"], "epoch": epoch},
+                    max_keys=["epoch"],
```
JSYK the checkpointer will always save the most recent checkpoints, the max_keys is not needed here, though its fine to have it to make things explicit.

pplantinga commented on speechbrain/lobes/models/kmeans.py:
```diff
@@ -0,0 +1,168 @@
+
+import joblib
+import torch
+from sklearn.cluster import MiniBatchKMeans
```
It would be cool to have this also in PyTorch, like this https://github.com/jokofa/torch_kmeans

But for practical reasons I guess it makes sense to keep it this way, though we will want to add it to `integrations` most likely.

lucadellalib replied:
Public PyTorch implementations typically implement K-means instead of mini-batch K-means, this requires several modifications. I had a PR for that (https://github.com/speechbrain/speechbrain/pull/2411), but it was decided to move on with the scikit-learn implementation. It would be hard now to switch while maintaining backward compatibility, since we have a lot of models trained using the scikit-learn recipe.

pplantinga commented on speechbrain/lobes/models/kmeans.py:
Should we deprecate the `speechbrain/utils/kmeans.py` in favor of this one?

pplantinga commented on recipes/LibriSpeech/quantization/train_kmeans.py:
```diff
@@ -0,0 +1,258 @@
+            self.feats = [self.feats[self.hparams.kmeans_batch_size :]]
+            self.curr_batch_size = len(self.feats[0])
+
+        if stage == sb.Stage.TRAIN:
```
Minor comment: there are two `if stage == train` in a row
```suggestion
```

pplantinga commented on recipes/LibriSpeech/quantization/train_kmeans.py:
```diff
@@ -0,0 +1,258 @@
+            # For K-means the validation/test loss is the inertia
+            # The lower the inertia, the better should be the clustering
+            # It is useful to monitor progress across epochs
+            # However, when saving checkpoints we always keep the last one (i.e. max_keys=["epoch"])
```
I don't suppose anyone has tested if clusters with lower inertia provide better quantization, i.e. better performance on downstream tasks...

lucadellalib replied:
Not that I'm aware of

pplantinga reviewed (commented):
Everything looks good, only minor comments.

lucadellalib commented:
@TParcollet @mravanelli @Adel-Moumen @pplantinga I've addressed the comments. I agree that moving this to `integration` makes sense. I compared my implementation to the original code, and when run with the same parameters and implementation details, it produces identical results. However, I made two modifications to the original code:
1) **Accumulated features**:  It looks like for some reason, in the original code, the feature list is emptied instead of keeping the remaining accumulated features for the next iteration (https://github.com/speechbrain/speechbrain/blob/v1.0.2/speechbrain/utils/kmeans.py#L197) 
2) **K-means batch size**: I believe the batch size is set incorrectly (`batch_size=hparams["batch_size"]` instead of `batch_size=hparams["kmeans_batch_size"]`), not sure what's the impact (https://github.com/speechbrain/speechbrain/blob/v1.0.2/recipes/LJSpeech/quantization/train.py#L119)

@poonehmousavi any insights?

mravanelli commented:
I ran the recipes again and everything works as expected. I think we can merge it, but I would like to give @TParcollet and @Adel-Moumen the opportunity for the last comments (if any).

Adel-Moumen commented on recipes/LJSpeech/quantization/hparams/train_discrete_ssl.yaml:
```diff
@@ -1,78 +1,102 @@
-    batch_size: !ref <batch_size>
-    drop_last: True
+# Preprocessing parameters
+# For some reason the duration is not available in the manifest file...
```
abt this, can't we simply modify the prepare recipe to have an extra duration column ?

lucadellalib replied:
Done

Adel-Moumen commented on recipes/LJSpeech/quantization/train.py:
```diff
@@ -1,142 +1,268 @@
-    for dataset in hparams["splits"]:
-        datasets[dataset] = sb.dataio.dataset.DynamicItemDataset.from_json(
-            json_path=data_info[dataset],
-            replacements={"data_root": hparams["data_folder"]},
-            dynamic_items=[audio_pipeline],
-            output_keys=["id", "sig"],
+        sig = torchaudio.functional.resample(
+            sig, original_sample_rate, hparams["sample_rate"]
         )
+        yield sig
```
why are you using this function with yield instead of the typical return?

lucadellalib replied:
I typically use `yield` in pipelines so they can be readily extended to return multiple values if needed

Adel-Moumen commented on recipes/LJSpeech/quantization/README.md:
```diff
@@ -1,74 +1,102 @@
+  pages   = {1--11},
+  url     = {http://jmlr.org/papers/v25/24-0991.html}
 }
-@misc{speechbrain,
-  title={{SpeechBrain}: A General-Purpose Speech Toolkit},
-  author={Mirco Ravanelli and Titouan Parcollet and Peter Plantinga and Aku Rouhe and Samuele Cornell and Loren Lugosch and Cem Subakan and Nauman Dawalatabad and Abdelwahab Heba and Jianyuan Zhong and Ju-Chieh Chou and Sung-Lin Yeh and Szu-Wei Fu and Chien-Feng Liao and Elena Rastorgueva and François Grondin and William Aris and Hwidong Na and Yan Gao and Renato De Mori and Yoshua Bengio},
-  year={2021},
-  eprint={2106.04624},
-  archivePrefix={arXiv},
-  primaryClass={eess.AS},
-  note={arXiv:2106.04624}
```
why did you replace this one?

Adel-Moumen commented on recipes/LJSpeech/quantization/README.md:
```diff
@@ -1,74 +1,102 @@
-  archivePrefix={arXiv},
-  primaryClass={cs.LG},
-  url={https://arxiv.org/abs/2407.00463},
+@article{ravanelli2024open,
+  author  = {Mirco Ravanelli and Titouan Parcollet and Adel Moumen and Sylvain de Langen and Cem Subakan and Peter Plantinga and Yingzhi Wang and Pooneh Mousavi and Luca {Della Libera} and Artem Ploujnikov and Francesco Paissan and Davide Borra and Salah Zaiem and Zeyu Zhao and Shucong Zhang and Georgios Karakasidis and Sung-Lin Yeh and Pierre Champion and Aku Rouhe and Rudolf Braun and Florian M ...
+  title   = {Open-Source Conversational {AI} with {SpeechBrain} 1.0},
+  journal = {Journal of Machine Learning Research},
+  year    = {2024},
+  volume  = {25},
+  number  = {333},
+  pages   = {1--11},
+  url     = {http://jmlr.org/papers/v25/24-0991.html}
 }
```
I understnad that this is the official new citation, but you could please replace `ravanelli2024open` by `speechbrainV1` to remain consistent with our past formatting. I will do the modifications for the other ones.

Adel-Moumen commented on recipes/LJSpeech/quantization/train.py:
```diff
@@ -1,142 +1,267 @@
+        curr_cluster_centers = self.hparams.quantizer.cluster_centers
+        loss = (curr_cluster_centers - old_cluster_centers).norm()
+
+        # Keep compatibility with standard supervised training
+        # (SpeechBrain expects a tensor with gradient)
+        loss.requires_grad_()
+        self.optimizer_step += 1
+        assert self.optimizer_step == self.modules.quantizer.n_steps, (
+            f"optimizer_step: {self.optimizer_step}",
+            f"quantizer.n_steps: {self.modules.quantizer.n_steps}",
+        )
```
remark: I actually do think that SpeechBrain Brain class will need at some point a small refactor to allow the possibility to return a Loss object that is not from PyTorch and where you won't do BP on it.

Adel-Moumen reviewed (approved):
LGTM. Thanks @lucadellalib for the refactoring.
